legate/core/launcher.py

```diff
@@ -92,84 +126,123 @@ def pack(self, buf):
 
         _pack(buf, self._value, dtype, is_tuple)
 
-    def __str__(self):
+    def __str__(self) -> str:
         return f"ScalarArg({self._value}, {self._dtype}, {self._untyped})"
 
-    def __repr__(self):
+    def __repr__(self) -> str:
         return str(self)
 
 
 class FutureStoreArg:
-    def __init__(self, store, read_only, has_storage, redop):
+    def __init__(
+        self, store: Store, read_only: bool, has_storage: bool, redop: int
+    ) -> None:
         self._store = store
         self._read_only = read_only
         self._has_storage = has_storage
         self._redop = redop
 
-    def pack(self, buf):
+    def pack(self, buf: BufferBuilder) -> None:
         self._store.serialize(buf)
         buf.pack_32bit_int(self._redop)
         buf.pack_bool(self._read_only)
         buf.pack_bool(self._has_storage)
         buf.pack_32bit_int(self._store.type.size)
         _pack(buf, self._store.extents, ty.int64, True)
 
-    def __str__(self):
+    def __str__(self) -> str:
         return f"FutureStoreArg({self._store})"
 
-    def __repr__(self):
+    def __repr__(self) -> str:
         return str(self)
 
 
 class RegionFieldArg:
-    def __init__(self, analyzer, store, dim, req, field_id, redop):
+    @overload
+    def __init__(
+        self,
+        analyzer: RequirementAnalyzer,
+        store: Store,
+        dim: int,
+        req: RegionReq,
+        field_id: int,
+        redop: int,
+    ) -> None:
+        ...
+
+    @overload
+    def __init__(
+        self,
+        analyzer: OutputAnalyzer,
+        store: Store,
+        dim: int,
+        req: OutputReq,
+        field_id: int,
+        redop: int,
+    ) -> None:
+        ...
+
+    def __init__(
+        self,
+        analyzer: Union[OutputAnalyzer, RequirementAnalyzer],
+        store: Store,
+        dim: int,
+        req: Union[OutputReq, RegionReq],
+        field_id: int,
+        redop: int,
+    ) -> None:
         self._analyzer = analyzer
         self._store = store
         self._dim = dim
         self._req = req
         self._field_id = field_id
         self._redop = redop
 
-    def pack(self, buf):
+    def pack(self, buf: BufferBuilder) -> None:
         self._store.serialize(buf)
         buf.pack_32bit_int(self._redop)
         buf.pack_32bit_int(self._dim)
+
         buf.pack_32bit_uint(
-            self._analyzer.get_requirement_index(self._req, self._field_id)
+            self._analyzer.get_requirement_index(
+                self._req, self._field_id  # type: ignore [arg-type]
+            )
         )
         buf.pack_32bit_uint(self._field_id)
 
-    def __str__(self):
+    def __str__(self) -> str:
         return f"RegionFieldArg({self._dim}, {self._req}, {self._field_id})"
 
-    def __repr__(self):
+    def __repr__(self) -> str:
         return str(self)
 
 
-_single_task_calls = {
+LegionTaskMethod = Any
```

Comment:
I wasn't able to get anything improved for this at present (I thought a callback protocol would suffice but it didn't). Refactoring this should probably be a dedicated PR.